source/fuzz/transformation_inline_function.cpp

```diff
@@ -0,0 +1,176 @@
+      ir_context->get_instr_block(function_call_instruction)->GetParent();
+  auto called_function = fuzzerutil::FindFunction(
+      ir_context, function_call_instruction->GetSingleWordInOperand(0));
+
+  for (auto& block : *called_function) {
+    // The called function entry block label will not be inlined.
+    if (&block != &*called_function->entry()) {
+      auto cloned_label_instruction = block.GetLabelInst()->Clone(ir_context);
+      uint32_t fresh_id =
+          message_.result_id_map().at(cloned_label_instruction->result_id());
+      cloned_label_instruction->SetResultId(fresh_id);
+      function_call_instruction->InsertBefore(
+          std::unique_ptr<opt::Instruction>(cloned_label_instruction));
+      fuzzerutil::UpdateModuleIdBound(ir_context, fresh_id);
+    }
+
+    for (auto& instruction : block) {
+      // The function return instruction will not be inlined.
+      if (&instruction == &*called_function->tail()->tail()) {
+        continue;
+      }
+
+      // Replaces the operand ids with their mapped result ids.
+      auto cloned_instruction = instruction.Clone(ir_context);
+      cloned_instruction->ForEachInId(
+          [this, called_function, function_call_instruction](uint32_t* id) {
+            // If the id is mapped, then set it to its mapped value.
+            if (message_.result_id_map().count(*id)) {
+              *id = message_.result_id_map().at(*id);
+              return;
+            }
+
+            uint32_t parameter_index = 0;
+            called_function->ForEachParam(
+                [id, function_call_instruction,
+                 &parameter_index](opt::Instruction* parameter_instruction) {
+                  // If the id is a function parameter, then set it to the
+                  // parameter value passed in the function call instruction.
+                  if (*id == parameter_instruction->result_id()) {
+                    *id = function_call_instruction->GetSingleWordInOperand(
+                        parameter_index + 1);
+                  }
+                  parameter_index++;
+                });
+          });
+
+      // If |cloned_instruction| has a result id, then set it to its mapped
+      // value.
+      if (cloned_instruction->HasResultId() &&
+          message_.result_id_map().count(cloned_instruction->result_id())) {
+        uint32_t result_id =
+            message_.result_id_map().at(cloned_instruction->result_id());
+        cloned_instruction->SetResultId(result_id);
+        fuzzerutil::UpdateModuleIdBound(ir_context, result_id);
+      }
+
+      // If the called function return instruction is OpReturnValue and the
+      // |cloned_instruction| result id is the returned value, then sets the
+      // |cloned_instruction| result id to the function call result id.
+      if (cloned_instruction->result_id() ==
+          called_function->GetReturnValueId()) {
+        cloned_instruction->SetResultId(function_call_instruction->result_id());
+      }
```

Comment:
This code should be moved to inside the "if (cloned_instruction->HasResultId())" above, otherwise you end up inspecting the result ids of instructions that don't have them and they may be spurious.

Reply:
You're right.